packages/node-core/src/indexer/dictionary/dictionary.service.ts

```diff
@@ -0,0 +1,144 @@
+      }
+    }
+  }
+
+  // Find the next valid dictionary
+  private findDictionary(height: number, skipDictionaryIndex: Set<number>) {
+    // DO NOT remove dictionary not valid
+    // As they can be valid for different block range, or work for other query
+    const index = this._dictionaries.findIndex(
+      (d, index) => d.heightValidation(height) && !skipDictionaryIndex.has(index)
+    );
+    // If not find any, then still set undefined.
+    this._currentDictionaryIndex = index < 0 ? undefined : index;
+  }
+
+  useDictionary(height: number): boolean {
+    return !!this.getDictionary(height);
+  }
+
+  /**
+   *
+   * @param dataSources
+   */
+
+  buildDictionaryEntryMap(dataSources: BlockHeightMap<DS[]>): void {
+    for (const dict of this._dictionaries) {
+      dict.updateQueriesMap(dataSources);
+    }
+  }
+
+  async scopedDictionaryEntries(
+    startBlockHeight: number,
+    scaledBatchSize: number,
+    latestFinalizedHeight: number, //api FinalizedHeight
+    skipDictionaryIndex: Set<number> = new Set<number>()
+  ): Promise<DictionaryResponse<number | IBlock<FB>> | undefined> {
+    // Initialize skipDictionaryIndex as an empty array
+    // Attempt to get data from the current dictionary
+    // const result = await this.tryGetDictionaryData(startBlockHeight, scaledBatchSize, latestFinalizedHeight, skipDictionaryIndex);
+    let result: DictionaryResponse<number | IBlock<FB>> | undefined;
+    const dictionary = this.getDictionary(startBlockHeight, skipDictionaryIndex);
+    if (!dictionary) {
+      return undefined;
+    }
+    try {
+      const queryEndBlock = dictionary.getQueryEndBlock(startBlockHeight + scaledBatchSize, latestFinalizedHeight);
+      result = await dictionary.getData(startBlockHeight, queryEndBlock, scaledBatchSize);
+    } catch (error: any) {
+      // Handle errors by skipping the current dictionary
+      assert(
+        this._currentDictionaryIndex !== undefined,
+        new Error(`try get next dictionary but _currentDictionaryIndex is undefined`)
+      );
+      skipDictionaryIndex.add(this._currentDictionaryIndex);
+      return this.scopedDictionaryEntries(
+        startBlockHeight,
+        scaledBatchSize,
+        latestFinalizedHeight,
+        skipDictionaryIndex
+      );
+    }
+    return result;
+  }
+
+  protected async resolveDictionary(
+    networkFamily: NETWORK_FAMILY,
+    chainId: string,
+    registryUrl: string
+  ): Promise<string | undefined> {
+    try {
+      const response = await fetch(registryUrl);
+
+      if (!response.ok) {
+        throw new Error(`Bad response, code="${response.status}" body="${await response.text()}"`);
+      }
+      const dictionaryJson = await response.json();
+
+      const dictionaries = dictionaryJson[networkFamily.toLowerCase()][chainId];
+
+      if (Array.isArray(dictionaries) && dictionaries.length > 0) {
+        // TODO choose alternatives
+        return dictionaries[0];
```

Comment:
This could be updated now so that we can create the appropriate v1 and v2 dictionaries